onnxruntime/core/providers/cpu/rnn/rnn_helpers.h

```diff
@@ -230,17 +230,51 @@ void ExecuteLambdaInParallel(const std::string& name, TLambda lambda, int max, i
   ORT_UNUSED_PARAMETER(name);
   ORT_UNUSED_PARAMETER(logger);
 
-  std::atomic<int> done(0);
-  for (int i = 0; i < max; i += step) {
-    ttp.Schedule([lambda, i, &done]() {
-      lambda(i);
-      ++done;
+  const int total_tasks = max / (step > 0 ? step : 1) + (max % step > 0 ? 1 : 0);
+
+  // ORT_ENFORCE may and does throw at times from within the tasks that run
+  // on a thread-pool. Without propagating exceptions the process exits silently
+  // which will make diagnosing bugs more difficult.
+
+  // \! UGLY
+  // We have a problem here with the current thread-pool is that it takes std::function
+  // by value and copies it more than once (even though it is movable).
+  //
+  // To report status and exceptions properly it's better to use
+  // futures and promises but they are not copyable, so we can't come up with a functor
+  // with a promise member and we are downgrading to C++11 where we can't have captures that moved in.
+  //
+  // At the same time promises MUST live in the child thread so if we throw from the main thread
+  // we don't destroy any promises that are on the main thread stack which children threads may still be using.
+  //
+  // The only solution with the current Eigen that comes to mind is to have shared_ptr to with std::promise.
+  //
+
+  std::vector<std::future<void> > futures;
+  for (int i = 0, t = 0; i < max; i += step, ++t) {
+    auto p_ptr = std::make_shared<std::promise<void> >();
+    futures.push_back(p_ptr->get_future());
+    ttp.Schedule([p_ptr, lambda, i]() {
+      try {
+        lambda(i);
+        p_ptr->set_value();
+      } catch (...) {
+        p_ptr->set_exception(std::current_exception());
+      }
     });
   }
 
-  int totalTasks = max / (step > 0 ? step : 1) + (max % step > 0 ? 1 : 0);
-  while (done != totalTasks)
-    ;
+  // If ORT_ENFORCE is failing multiple threads can throw
+  // we ignore other exceptions but one. Futures with unclaimed
+  // shared states should destruct just fine.
+  for (auto& fut : futures) {
```

Comment:
It's better but not the best. I don't like the C++ futures and promises because you can't wait on multiple futures at the same time. Which means, even one of the tasks already failed, you still need to wait all the other to be finished.  If the language can provide a way like the WaitForMultipleObjects, then the error could be propagated faster.

Reply:
Are we concerned about how performant failures are?

---
In reply to: [327840287](https://github.com/microsoft/onnxruntime/pull/1905#discussion_r327840287) [](ancestors = 327840287)